### DISCUSSION TITLE
VueNodeViewRenderer is triggered only once, and returns empty object

### DISCUSSION THREAD

**Discussion by jvissers** [ueberdosis/tiptap#2388] [General Questions] [answered]
I have the following arrangement:

![metadata-editor](https://user-images.githubusercontent.com/2955276/149509634-8c565c2f-fc83-4dc0-a558-b6f970297a7a.png)

In words. My goal is to have an editor shell, capable of loading a collection of external (tiptap) extensions, based on a piece of metadata that gets fetched. Along with that metadata the fetch data also contains the content for the editor to be displayed. The external extensions are loaded via **dynamic imports** and added to an array of _AnyExtension[]_. Remote extensions might be very simple, e.g. wrappers around default tiptap extensions, document, paragraph, text or more involved. One example would be an extensions that uses Vue3 nodeview. All extensions are built as a seperate package, using vitejs **library mode** each results in an ES6 module. Foundational modules that are uses throughout all extensions, like `@tiptap/core` but also `@tiptap/vue-3` are tagged as being external - so these won't be bundled in the ES6 module. Instead these modules are loaded through _import maps_ and exposed that way. Right now this all works fine, expect for components that have Vue3 nodeviews. For these nodes, the component tags are not processed and effectively end-up in the HTML as-is (e.g. `<vue-counter count="10"></vue-counter>`). This seems to be related to the fact that the `VueNodeViewRenderer` function is only invoked once, where I think it needs to be invoked twice for each custom node that uses Vue3 nodeviews (at least that is what I'm seeing in an example that doesn't use importmaps, nor dynamic imports.

Caveat. The main caveat with my setup is that because of the fact that dynamic imports, importmaps do not play nice with module bundlers - I've resorted to a setup of only typescript and live-server. This means that the **EditorShell** cannot use a SFC .vue file. Instead I'm defining this as a SFC through directly going to the composition API. Right now I have this:

```
import { EditorContent, useEditor } from '@tiptap/vue-3';
import { AnyExtension, Content } from '@tiptap/core';
import { h, defineComponent } from 'vue';

import useContentModel from '../utils/useContentModel.js';

const { showContentModel } = useContentModel();

const EditorShell = defineComponent({
  props: {
    element: {},
    extensions: {},
    content: {},
  },

  setup(props) {
    console.log('Calling setup...', props);
    const editor = useEditor({
      extensions: props.extensions as AnyExtension[],
      element: props.element as HTMLElement,
      content: props.content as Content,

      onCreate({ editor }) {
        console.log('Document schema: ', JSON.stringify(editor.schema.spec, null, 2));
        console.log('Loaded document valid =', editor.state.doc.check() === undefined);
        showContentModel(editor);
      },

      onUpdate({ editor }) {
        showContentModel(editor);
      },
    });

    return { editor };
  },

  render() {
    console.log('Calling render...', this.editor);
    return h(EditorContent, {
      editor: this.editor,
    });
  },
});

export default EditorShell;
```

I'm invoking this _EditorShell_ from my 'main' as follows:

```
     const shell = document.querySelector<HTMLDivElement>('#editor')!;
      const editorShell = createApp({
        render: () => h(EditorShell),
      }, { "extensions": extensions, "element": shell, "content": content});
      editorShell.mount('#editor');
```

In the above:

- extensions ~ is the collection of remote extensions that is dynamically imported/resolved
- shell ~ is the div in the HTML that must hold the editor shell.
- content ~ is the actual content that is loaded.

With this I'm seeing that the `console.log` statements as are specified on the `onCreate()` method above, echo the correct information. However the `render()` operation basically has no effect. The method `render()` is indeed called twice, judging from the log.

- Calling render... undefined EditorShell.ts:39 
- Calling render... Editor {callbacks: {…}, isFocused: false, extensionStorage: {…}, options: {…}, isCapturingTransaction: false, …}...

However, again, VueNodeViewRenderer is only invoked once.

**Question: what could be possible causes of this happening?**

**Comment by jvissers** [comment]
After noticing that in the above setup, still some parts of the editor got rendered - I rewrote the `EditorShell` to more closely resemble the example that is on the tiptap website. More specifically I'm now creating the editor in `mounted()` instead of `setup()`, which I think should work. However, still not seeing the correct result - as `render()` effectively doesn't render anything.

```
import { Editor, EditorContent } from '@tiptap/vue-3';
import { AnyExtension, Content } from '@tiptap/core';
import { h, defineComponent } from 'vue';

import useContentModel from '../utils/useContentModel.js';

const { showContentModel } = useContentModel();

const EditorShell = defineComponent({
  props: {
    element: {},
    extensions: {},
    content: {},
  },

  data() {
    return {
      editor: null,
    };
  },

  mounted() {
    this.editor = new Editor({
      extensions: this.extensions as AnyExtension[],
      element: this.element as HTMLElement,
      content: this.content as Content,

      onCreate({ editor }) {
        console.log('Document schema: ', JSON.stringify(editor.schema.spec, null, 2));
        console.log('Loaded document valid =', editor.state.doc.check() === undefined);
        showContentModel(editor);
      },

      onUpdate({ editor }) {
        showContentModel(editor);
      },
    });
  },

  beforeUnmount() {
    this.editor.destroy();
  },

  render() {
    return h(EditorContent, {
      editor: this.editor,
    });
  },
});

export default EditorShell;
```

So what should that `render()` function look like, if I cannot use a .vue file?

**Comment by jvissers** [comment]
The above change - did improve the situation albeit marginally, in that while debugging I can now see almost the same behavior of my arrangement and the default example as is described here: [Render a Vue component](https://tiptap.dev/guide/node-views/vue#render-a-vue-component), which is:

- the view is rendered a first time, the custom Vue3 nodeview isn't visible
- the view is rendered a second time, the Vue3 nodeview is now visible

The main difference in my specific case is that during the 2nd render, effectively everything that was put there the 1st time around will vanish after that 2nd render.

**Comment by jvissers** [comment]
Moving up-and-down the stack noticed yet another difference, which could be related. Debugging the [Render a Vue component](https://tiptap.dev/guide/node-views/vue#render-a-vue-component), I'm noticing that this code of the custom tiptap node:

```
  addNodeView() {
    return VueNodeViewRenderer(CounterComponent);
  },
```

is called twice. However this is not the case in the code that works with remote extensions. In that situation there is only one call to `addNodeView()`. 

**Question: Could this be related?**

**Comment by jvissers** [comment]
Seeing this error: **The new child element contains the parent.** 

![DOM-error](https://user-images.githubusercontent.com/2955276/149538341-ab7704eb-d6d8-4f3f-a30f-88e2d243b766.png)

in `EditorContent.ts` - this seems to interrupt the creation of editor's nodeviews in this tiptap code:

```
watchEffect(() => {
      const editor = props.editor

      if (editor && editor.options.element && rootEl.value) {
        nextTick(() => {
          if (!rootEl.value || !editor.options.element.firstChild) {
            return
          }

          const element = unref(rootEl.value)

          rootEl.value.append(...editor.options.element.childNodes)

          // @ts-ignore
          editor.contentComponent = instance.ctx._

          editor.setOptions({
            element,
          })

          editor.createNodeViews()
        })
      }
    })
```

I must be missing something, as right now I don't have a clue how to control this.

**Comment by jvissers** [comment]
I'm pretty sure that the above exception: `The new child element contains the parent.` is the thing that needs to be solved first. From the looks of it, I'm not targetting the correct div somehow.

**Comment by jvissers** [comment, marked as answer]
Found the issue.

My `EditorShell` had this mount function:

```
mounted() {
    this.editor = new Editor({
      extensions: this.extensions as AnyExtension[],
      element: this.element as HTMLElement,
      content: this.content as Content,

      onCreate({ editor }) {
        console.log('Document schema: ', JSON.stringify(editor.schema.spec, null, 2));
        console.log('Loaded document valid =', editor.state.doc.check() === undefined);
        showContentModel(editor);
      },

      onUpdate({ editor }) {
        showContentModel(editor);
      },
    });
  },
```

And I was calling this from my main like so:

```
      const shell = document.querySelector<HTMLDivElement>('#editor')!;
      // prettier-ignore
      const editorShell = createApp({
        render: () => h(EditorShell),
      }, { "extensions": extensions, "element": shell, "content": content});
      editorShell.mount('#editor');
```  

Effectively what needed to be changed was to remove the `element` property from the `EditorShell`, and no longer pass that as a property. Instead the `editorShell.mount('#editor');` is enough and resolves `The new child element contains the parent` issue.

**Reply by hanspagel** [reply]
Thanks for sharing! 🙃
